File: app/app/src/main/java/com/example/hakonsreader/viewmodels/SearchForSubredditsViewModel.java
package com.example.hakonsreader.viewmodels;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.hakonsreader.App;
import com.example.hakonsreader.api.RedditApi;
import com.example.hakonsreader.api.model.Subreddit;

import java.util.ArrayList;
import java.util.List;

/**
 * ViewModel for searching for subreddits
 */
public class SearchForSubredditsViewModel extends ViewModel {

    private final RedditApi api = App.get().getApi();
    private final MutableLiveData<List<Subreddit>> searchResults = new MutableLiveData<>();
    private final MutableLiveData<ErrorWrapper> error = new MutableLiveData<>();
    private final MutableLiveData<Boolean> onCountChange = new MutableLiveData<>();


    /**
     * List of subreddits searched for
     *
     * @return A LiveData that can be observed to be notified of search results
     */
    public LiveData<List<Subreddit>> getSearchResults() {
        return searchResults;
    }

    /**
     * Errors received by the API
     *
     * @return A LiveData that can be observed to be notified when API errors occur
     */
    public LiveData<ErrorWrapper> getError() {
        return error;
    }

    public LiveData<Boolean> getOnCountChange() {
        return onCountChange;
    }

    /**
     * Searches for subreddits
     *
     * @param query The search query
     */
    public void search(String query) {
        onCountChange.postValue(true);
        api.subreddits().search(query, subreddits -> {
            onCountChange.postValue(false);
            searchResults.postValue(subreddits);
        }, (e, t) -> {
            onCountChange.postValue(false);
            error.postValue(new ErrorWrapper(e, t));
        });
    }

    /**
     * Clears the list of subreddits searched for
     */
    public void clearSearchResults() {
        searchResults.postValue(new ArrayList<>());
    }
}

File: app/app/src/main/java/com/example/hakonsreader/viewmodels/SearchForSubredditsViewModel.kt
package com.example.hakonsreader.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.hakonsreader.App
import com.example.hakonsreader.api.model.Subreddit
import com.example.hakonsreader.api.responses.ApiResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.launch

class SearchForSubredditsViewModel : ViewModel() {

    private val api = App.get().api
    private val _searchResults = MutableLiveData<List<Subreddit>>()
    private val _isLoading = MutableLiveData<Boolean>()
    private val _error = MutableLiveData<ErrorWrapper>()

    /**
     * Map containing the previous search results, mapping a search query to a list of subreddits
     */
    private val cachedSearchResults = HashMap<String, List<Subreddit>>()

    val searchResults : LiveData<List<Subreddit>> = _searchResults
    val isLoading: LiveData<Boolean> = _isLoading
    val error: LiveData<ErrorWrapper> = _error


    /**
     * Search for subreddits
     *
     * @param query The search query
     */
    fun search(query: String) {
        if (cachedSearchResults.containsKey(query)) {
            _searchResults.postValue(cachedSearchResults[query])
            return
        }

        _isLoading.postValue(true)

        viewModelScope.launch {
            val resp = api.subreditts().search(query)
            _isLoading.postValue(false)

            when (resp) {
                is ApiResponse.Success -> {
                    cachedSearchResults[query] = resp.value
                    _searchResults.postValue(resp.value!!)
                }
                is ApiResponse.Error -> _error.postValue(ErrorWrapper(resp.error, resp.throwable))
            }
        }
    }

    /**
     * Clear the search results
     */
    fun clearSearchResults() {
        _searchResults.postValue(ArrayList())
    }
}